toolchain/parse/handle_var.cpp

```diff
@@ -6,22 +6,41 @@
 
 namespace Carbon::Parse {
 
-// Handles VarAs(Semicolon|For).
+// Handles VarAs(Declaration|For).
 static auto HandleVar(Context& context, State finish_state) -> void {
   context.PopAndDiscardState();
 
-  // These will start at the `var`.
+  // These will start at the `var` or `returned`.
   context.PushState(finish_state);
   context.PushState(State::VarAfterPattern);
 
+  // Consume the `var` or `returned var`.
+  auto returned_token = context.ConsumeIf(Lex::TokenKind::Returned);
+  if (returned_token && !context.PositionIs(Lex::TokenKind::Var)) {
+    CARBON_DIAGNOSTIC(ExpectedVarAfterReturned, Error,
+                      "Expected `var` after `returned`.");
+    context.emitter().Emit(*context.position(), ExpectedVarAfterReturned);
+    context.PopAndDiscardState();
+    context.PopAndDiscardState();
```

Comment:
It looks a bit like you're pushing state that may immediately be discarded. Can the order be switched?

Reply:
Done. It's slightly awkward because we need to enter these states before we create the introducer which ends up being a child of the nodes created for them, but want to parse the introducer before we decide whether to enter the states.